File: OsmAndWear/src/net/osmand/plus/helpers/GmsLocationServiceHelper.java
package net.osmand.plus.helpers;

import static android.content.Context.LOCATION_SERVICE;

import android.location.Location;
import android.location.LocationManager;
import android.os.Looper;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.location.LocationManagerCompat;
import androidx.core.location.LocationRequestCompat;

import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationAvailability;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.location.Priority;
import com.google.android.gms.tasks.Task;

import net.osmand.PlatformUtil;
import net.osmand.plus.OsmandApplication;

import org.apache.commons.logging.Log;

import java.util.Collections;
import java.util.List;

public class GmsLocationServiceHelper extends LocationServiceHelper {

	private static final Log LOG = PlatformUtil.getLog(DayNightHelper.class);

	// FusedLocationProviderClient - Main class for receiving location updates.
	private final FusedLocationProviderClient fusedLocationProviderClient;

	// LocationRequest - Requirements for the location updates, i.e., how often you should receive
	// updates, the priority, etc.
	private final LocationRequest fusedLocationRequest;
	private final LocationRequestCompat networkLocationRequest;

	// LocationCallback - Called when FusedLocationProviderClient has a new Location.
	private final com.google.android.gms.location.LocationCallback fusedLocationCallback;

	private LocationCallback locationCallback;

	public GmsLocationServiceHelper(@NonNull OsmandApplication app) {
		super(app);

		fusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(app);
		fusedLocationRequest = new LocationRequest.Builder(Priority.PRIORITY_HIGH_ACCURACY, 100).build();
		networkLocationRequest = new LocationRequestCompat.Builder(100).setQuality(LocationRequestCompat.QUALITY_HIGH_ACCURACY).build();
		fusedLocationCallback = new com.google.android.gms.location.LocationCallback() {
			@Override
			public void onLocationResult(@NonNull LocationResult locationResult) {
				LocationCallback locationCallback = GmsLocationServiceHelper.this.locationCallback;
				if (locationCallback != null) {
					Location location = locationResult.getLastLocation();
					net.osmand.Location l = convertLocation(location);
					locationCallback.onLocationResult(l == null ? Collections.emptyList() : Collections.singletonList(l));
				}
			}

			@Override
			public void onLocationAvailability(@NonNull LocationAvailability locationAvailability) {
				LocationCallback locationCallback = GmsLocationServiceHelper.this.locationCallback;
				if (locationCallback != null) {
					locationCallback.onLocationAvailability(locationAvailability.isLocationAvailable());
				}
			}
		};
	}

	@Override
	public void requestLocationUpdates(@NonNull LocationCallback locationCallback) {
		this.locationCallback = locationCallback;
		// request location updates
		try {
			fusedLocationProviderClient.requestLocationUpdates(
					fusedLocationRequest, fusedLocationCallback, Looper.myLooper());
		} catch (SecurityException e) {
			LOG.debug("Location service permission not granted");
			throw e;
		} catch (IllegalArgumentException e) {
			LOG.debug("GPS location provider not available");
			throw e;
		}
	}

	@Override
	public boolean isNetworkLocationUpdatesSupported() {
		return true;
	}

	@Override
	public void requestNetworkLocationUpdates(@NonNull LocationCallback locationCallback) {
		this.networkLocationCallback = locationCallback;
		// request location updates
		LocationManager locationManager = (LocationManager) app.getSystemService(LOCATION_SERVICE);
		List<String> providers = locationManager.getProviders(true);
		for (String provider : providers) {
			if (provider == null
					|| provider.equals(LocationManager.GPS_PROVIDER)
					|| provider.equals(LocationManager.FUSED_PROVIDER)) {
				continue;
			}
			try {
				NetworkListener networkListener = new NetworkListener();
				LocationManagerCompat.requestLocationUpdates(locationManager, provider, networkLocationRequest,
						networkListener, Looper.myLooper());
				networkListeners.add(networkListener);
			} catch (SecurityException e) {
				LOG.debug(provider + " location service permission not granted");
			} catch (IllegalArgumentException e) {
				LOG.debug(provider + " location provider not available");
			}
		}
	}

	@Override
	public void removeLocationUpdates() {
		// remove location updates
		try {
			fusedLocationProviderClient.removeLocationUpdates(fusedLocationCallback);
		} catch (SecurityException e) {
			LOG.debug("Location service permission not granted", e);
			throw e;
		} finally {
			removeNetworkLocationUpdates();
		}
	}

	@Nullable
	public net.osmand.Location getFirstTimeRunDefaultLocation(@Nullable LocationCallback locationCallback) {
		if (locationCallback == null) {
			return null;
		}
		try {
			Task<Location> lastLocation = fusedLocationProviderClient.getLastLocation();
			lastLocation.addOnSuccessListener(loc -> locationCallback.onLocationResult(loc != null
					? Collections.singletonList(convertLocation(loc)) : Collections.emptyList() ));
		} catch (SecurityException e) {
			LOG.debug("Location service permission not granted");
		} catch (IllegalArgumentException e) {
			LOG.debug("GPS location provider not available");
		}
		return null;
	}
}

File: OsmAnd-telegram/src/net/osmand/telegram/helpers/location/GmsLocationServiceHelper.kt
package net.osmand.telegram.helpers.location

import android.os.Looper
import android.util.Log
import com.google.android.gms.location.FusedLocationProviderClient
import com.google.android.gms.location.LocationAvailability
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationResult
import com.google.android.gms.location.LocationServices
import com.google.android.gms.location.Priority
import net.osmand.Location
import net.osmand.PlatformUtil
import net.osmand.telegram.TelegramApplication
import net.osmand.telegram.utils.OsmandLocationUtils.convertLocation

class GmsLocationServiceHelper(app: TelegramApplication): LocationServiceHelper() {

	// FusedLocationProviderClient - Main class for receiving location updates.
	private val fusedLocationProviderClient : FusedLocationProviderClient

	// LocationRequest - Requirements for the location updates, i.e., how often you should receive
	// updates, the priority, etc.
	private val fusedLocationRequest : LocationRequest

	// Native LocationCallback from Google Services
	// Called when FusedLocationProviderClient has a new Location.
	private val fusedLocationCallback : com.google.android.gms.location.LocationCallback

	// Callback to return results of calling helper methods to external requesters.
	private var locationCallback: LocationCallback? = null

	init {
		fusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(app)

		// Sets the desired interval for active location updates. This interval is inexact. You
		// may not receive updates at all if no location sources are available, or you may
		// receive them less frequently than requested. You may also receive updates more
		// frequently than requested if other applications are requesting location at a more
		// frequent interval.
		//
		// IMPORTANT NOTE: Apps running on Android 8.0 and higher devices (regardless of
		// targetSdkVersion) may receive updates less frequently than this interval when the app
		// is no longer in the foreground.
		val interval = 1000L

		fusedLocationRequest = LocationRequest.Builder(Priority.PRIORITY_HIGH_ACCURACY, interval).build()

		fusedLocationCallback = object  : com.google.android.gms.location.LocationCallback() {

			override fun onLocationResult(locationResult: LocationResult) {
				val lastLocation = locationResult.lastLocation
				val result: List<Location> = if (lastLocation != null) listOf(convertLocation(lastLocation)!!) else emptyList()
				locationCallback?.onLocationResult(result)
			}

			override fun onLocationAvailability(locationAvailability: LocationAvailability) {
				locationCallback?.onLocationAvailability(locationAvailability.isLocationAvailable)
			}
		}
	}

	override fun requestLocationUpdates(callback: LocationCallback) {
		this.locationCallback = callback
		try {
			fusedLocationProviderClient.requestLocationUpdates(
				fusedLocationRequest, fusedLocationCallback, Looper.myLooper())
		} catch (e: SecurityException) {
			Log.d(PlatformUtil.TAG, "Lost location permissions. Couldn't request updates. $e")
			throw e
		} catch (e: IllegalArgumentException) {
			Log.d(PlatformUtil.TAG, "GPS location provider not available")
			throw e
		}
	}

	override fun isNetworkLocationUpdatesSupported() = false

	override fun requestNetworkLocationUpdates(callback: LocationCallback) {}

	override fun removeLocationUpdates() {
		try {
			fusedLocationProviderClient.removeLocationUpdates(fusedLocationCallback)
		} catch (e: SecurityException) {
			Log.d(PlatformUtil.TAG, "Lost location permissions. Couldn't remove updates. $e")
			throw e
		}
	}

	override fun getFirstTimeRunDefaultLocation(callback: LocationCallback?): Location? {
		locationCallback ?: return null
		try {
			val lastLocation = fusedLocationProviderClient.lastLocation
			lastLocation.addOnSuccessListener { loc: android.location.Location? ->
				val result = if (loc != null) listOf(convertLocation(loc)!!) else emptyList()
				locationCallback!!.onLocationResult(result)
			}
		} catch (e: SecurityException) {
			Log.d(PlatformUtil.TAG, "Location service permission not granted")
		} catch (e: IllegalArgumentException) {
			Log.d(PlatformUtil.TAG, "GPS location provider not available")
		}
		return null
	}
}